### PR TITLE
CORE-17532: Upgrade Liquibase to latest version

### PR DESCRIPTION
**Pull request by malachyb**
The version of liquibase used in corda is currently out of date (4.19, released in January 2023). This PR upgrades to a more modern 4.24, which includes many security updates and performance upgrades.

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -67,7 +67,7 @@ kryoVersion = 5.5.0
 kryoSerializersVersion = 0.45
 kotlinCoroutinesVersion=1.6.4
 # Liquibase upgrade to 4.20 and 4.21 failed - see CORE-12612 for more details
```
Comment (driessamyn) [inline gradle.properties]: Please remove this comment.

File: gradle.properties [hunk 2]
```diff
@@ -67,7 +67,7 @@ kryoVersion = 5.5.0
...
-liquibaseVersion = 4.19.0
+liquibaseVersion = 4.24.0
```
Comment (driessamyn) [inline gradle.properties]: Can this be moved to the Version Catalog as part of this PR?

File: libs/db/db-admin-impl/src/integrationTest/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImplTest.kt [hunk 3]
```diff
@@ -76,23 +76,11 @@ class LiquibaseSchemaMigratorImplTest {
...
-        ds.connection.use { db ->
-            sql1.lines()
-                .filter {
-                    it.isNotBlank() && !it.startsWith("--")
-                }
-                .forEach {
-                    db.createStatement().execute(it)
-                }
-        }
+            .containsPattern("run: +3")
```
Comment (driessamyn) [inline libs/db/db-admin-impl/src/integrationTest/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImplTest.kt]: Why is this changed, and why is the test that executes the output SQL removed?
This diff suggests we no longer generate valid SQL

File: libs/db/db-admin-impl/src/integrationTest/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImplTest.kt [hunk 4]
```diff
@@ -101,21 +89,10 @@ class LiquibaseSchemaMigratorImplTest {
...
-        ds.connection.use { db ->
-            sql2.lines()
-                .filter {
-                    it.isNotBlank() && !it.startsWith("--")
-                }
-                .forEach {
-                    db.createStatement().execute(it)
-                }
-        }
+            .containsPattern("run: +1")
```
Comment (driessamyn) [inline libs/db/db-admin-impl/src/integrationTest/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImplTest.kt]: As above

File: libs/virtual-node/cpi-datamodel/build.gradle [hunk 5]
```diff
@@ -7,6 +7,7 @@ description 'CPI Data Model'
 
 dependencies {
     compileOnly 'org.osgi:osgi.annotation'
+    implementation "org.liquibase:liquibase-core:$liquibaseVersion"
```
Comment (driessamyn) [inline libs/virtual-node/cpi-datamodel/build.gradle]: Why is this needed?

**Review by vkolomeyko — DISMISSED** [review]
LGTM

File: libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt [hunk 6]
```diff
@@ -29,7 +42,8 @@ class LiquibaseSchemaMigratorImpl(
             DatabaseFactory
                 .getInstance()
                 .findCorrectDatabaseImplementation(JdbcConnection(connection))
-        }
+        },
+    private val runChanges: Boolean = true
```
Comment (driessamyn) [inline libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt]: Why do we need this new argument?

File: libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt [hunk 7]
```diff
@@ -110,12 +120,52 @@ class LiquibaseSchemaMigratorImpl(
...
+            )
+
+            try {
+                val scopeObjects = mapOf(
+                    Scope.Attr.database.name to lb.database,
+                    Scope.Attr.resourceAccessor.name to lb.resourceAccessor
+                )
+                Scope.child(scopeObjects) {
+                    val command: CommandScope
+                    if (sql != null) {
```
Comment (driessamyn) [inline libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt:137]: There are now 2 ways of identifying whether the SQL should be executed or not.

It may be simpler to do something along the lines of

```kotlin
val runChanges = null != sql

val command = if(runChanges) {
  CommandScope(UpdateCommandStep.COMMAND_NAME[0])
} else {
  CommandScope(UpdateSqlCommandStep.COMMAND_NAME[0]).also {
                        it.setOutput(
                            WriterOutputStream(
                                sql,
                                GlobalConfiguration.OUTPUT_FILE_ENCODING.currentValue
                            )
                        )
  }
}
```

File: libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt (hunk 6 above)
Comment (malachyb) [reply libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt]: This is for testing purposes (specifically libs/db/db-admin-impl/src/test/kotlin[...]/LiquibaseSchemaMigratorTest.kt) because mocking CommandScope caused it to need us to be able to import liquibase in a lot of places, so I instead added a toggle to prevent it from actually running the changes.

Comment (driessamyn) [reply libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt]: As discussed on Slack yesterday, I don't understand this comment.
See https://github.com/corda/corda-runtime-os/pull/5421 and in particular https://github.com/corda/corda-runtime-os/pull/5421/commits/a9b7d42d5e37cc6f9ecc15a0f465181c82dbf424 that shows the new argument is completely redundant.

File: libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt [hunk 8]
```diff
@@ -110,12 +123,39 @@ class LiquibaseSchemaMigratorImpl(
...
+                Scope.child(scopeObjects) {
+                    if (sql != null) {
+                        commandScopeFactory(UpdateSqlCommandStep.COMMAND_NAME).configure(lb, tag).also {
+                            it.setOutput(
+                                WriterOutputStream(
+                                    sql,
+                                    GlobalConfiguration.OUTPUT_FILE_ENCODING.currentValue
+                                )
+                            )
+                            it.execute()
```
Comment (driessamyn) [inline libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt:145]: I don't think you need this execute

Comment (malachyb) [reply libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt:145]: I do, the execute makes it actually generate the SQL. I tested without the execute and the SQL writer was empty.

**Comment by malachyb** [comment]
build e2e

**Comment by malachyb** [comment]
build e2e

**Comment by malachyb** [comment]
build e2e
